syntax = "proto3";

option java_multiple_files = true;
option java_package = "org.infinispan.grpc";
option java_outer_classname = "CacheProto";
option objc_class_prefix = "HLW";

service Cache {
  rpc topologyGetInfo (VoidMsg) returns (TopologyInfoMsg) {}
  rpc get (KeyMsg) returns (ValueMsg) {}
  rpc put (KeyValuePairMsg) returns (ValueMsg) {}
}

message ComplexKey {
    string name = 1;
    int32 domainId = 2;
}

message ComplexValue {
    string message = 1;
    float domainValue = 2;
}
message KeyMsg {
    oneof KeyMsgOneOf {
  string name = 1;
  ComplexKey complexName= 2;
    }
}

message ValueMsg {
    oneof ValueMsgOneOf {
  string message = 1;
  ComplexValue complexMessage = 2;
    }
}

message KeyValuePairMsg {
  KeyMsg key = 1;
  ValueMsg value = 2;
}

message VoidMsg {
}

message TopologyInfoMsg {
  repeated string serverId = 1;
}
